src/transformers/models/mistral/modeling_mistral.py

```diff
@@ -950,74 +936,36 @@ def forward(
         return_dict = return_dict if return_dict is not None else self.config.use_return_dict
 
         # retrieve input_ids and inputs_embeds
```

Comment:
all of that forward layer should get a copied from llama I think no?

Reply:
I think Mistral have this extra logic which I move inside `_update_causal_mask`:
```
 if self._attn_implementation == "flash_attention_2":
        if attention_mask is not None and use_cache:
                is_padding_right = attention_mask[:, -1].sum().item() != input_tensor.size()[0]
                if is_padding_right:
                    raise ValueError(
                        "You are attempting to perform batched generation with padding_side='right'"
                        " this may lead to unexpected behaviour for Flash Attention version of Mistral. Make sure to "
                        " call `tokenizer.padding_side  = 'left'` before tokenizing the input. "
                    )
        if attention_mask is not None and 0.0 in attention_mask:
                return attention_mask
        return None
```

Reply:
Got you!

Reply:
that's because we want to isolate the code changes as much as possible and standardize things when they are the same. 
For me the logic here looks very similar, but you are right, the forward cannot be copied from because it forces you to use some hacks in the update causal mask 😉

Reply:
copied form is also a lot easier to maintain and review